astropy/utils/misc.py

```diff
@@ -850,3 +850,134 @@ def set_locale(name):
+
+
+@six.add_metaclass(abc.ABCMeta)
+class ShapedLikeNDArray(object):
+    """Mixin class to provide shape-changing methods.
+
+    The class proper must define a ``shape`` property and ``_replicate`` method.
+    """
+
+    @abc.abstractproperty
+    def shape(self):
+        """The shape of the instance and underlying arrays."""
+
+    @abc.abstractmethod
+    def _replicate(method, *args, **kwargs):
+        """Replicate an instance, applying ``method`` to the underlying arrays.
+
+        Parameters
+        ----------
+        method : str
+            Method to be be applied to the internal arrays holding the instance
+            data (e.g., ``jd1`` and ``jd2`` for `~astropy.time.Time`)
+        args : tuple
+            Any positional arguments for ``method``.
+        kwargs : dict
+            Any keyword arguments for ``method``.
+        """
+
+    # Typically, ndim, size, and isscalar will be overridden, but we might as
+    # well set sensible defaults.
+    @property
+    def ndim(self):
+        """The number of dimensions of the instance and underlying arrays."""
+        return len(self.shape)
+
+    @property
+    def size(self):
+        """The size of the object, as calculated from its shape."""
+        size = 1
+        for sh in self.shape:
+            size *= sh
+        return size
+
+    @property
+    def isscalar(self):
+        return self.shape == ()
+
+    def __getitem__(self, item):
+        if self.isscalar:
+            raise TypeError('scalar {0!r} object is not subscriptable.'.format(
+                self.__class__.__name__))
+        return self._replicate('__getitem__', item)
+
+    def reshape(self, *args, **kwargs):
+        """Returns an instance containing the same data with a new shape.
+
+        Parameters are as for :meth:`~numpy.ndarray.reshape`.  Note that it is
+        not always possible to change the shape of an array without copying the
+        data. If you want an error to be raise if the data is copied, you
+        should assign the new shape to the shape attribute.
+        """
+        return self._replicate('reshape', *args, **kwargs)
+
+    def ravel(self, *args, **kwargs):
+        """Return an instance with the array collapsed into one dimension.
+
+        Parameters are as for :meth:`~numpy.ndarray.ravel`. Note that it is
+        not always possible to unravel an array without copying the data.
+        If you want an error to be raise if the data is copied, you should
+        should assign shape ``(-1,)`` to the shape attribute.
+        """
+        return self._replicate('ravel', *args, **kwargs)
+
+    def flatten(self, *args, **kwargs):
+        """Return a copy with the array collapsed into one dimension.
+
+        Parameters are as for :meth:`~numpy.ndarray.flatten`.
+        """
+        return self._replicate('flatten', *args, **kwargs)
+
+    def transpose(self, *args, **kwargs):
+        """Return an instance with the data transposed.
+
```

Comment:
behaviour is pretty complex for n-D arrays. Should we refer to the numpy documentation for more details?